benches/dapol.rs

```diff
@@ -1,175 +1,385 @@
-                build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height)
-            });
-        });
-    }
 
-    // bench tree height = 32
-    let tree_height = 32;
-    for &num_leaves in NUM_LEAVES.iter() {
-        let items = build_item_list(num_leaves, tree_height);
-        group.bench_function(BenchmarkId::new("height_32", num_leaves), |bench| {
-            bench.iter(|| {
-                // we bench range proof padding only because building a tree does not depend on
-                // the type of range proof we do
-                build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height)
-            });
-        });
-    }
+    dapol::initialize_machine_parallelism();
 
-    group.finish();
-}
+    let max_thread_count: u8 = MaxThreadCount::default().get_value();
 
-fn generate_proof(c: &mut Criterion) {
-    let mut group = c.benchmark_group("prove");
-    group.sample_size(10);
+    let mut thread_counts: Vec<u8> = Vec::new();
 
-    // this benchmark depends on the tree height and not the number of leaves,
-    // so we just pick the smallest number of leaves
-    let num_leaves = NUM_LEAVES[0];
-    for &tree_height in TREE_HEIGHTS.iter() {
-        let items = build_item_list(num_leaves, tree_height);
-        let mut rng = thread_rng();
-        let item_range = Uniform::new(0usize, num_leaves);
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofSplitting>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("splitting", tree_height), |bench| {
-            bench.iter(|| {
-                // time proof generation
-                let tree_index = &items[rng.sample(item_range)].0;
-                dapol.generate_proof(tree_index).unwrap()
-            });
-        });
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("padding", tree_height), |bench| {
-            bench.iter(|| {
-                // time proof generation
-                let tree_index = &items[rng.sample(item_range)].0;
-                dapol.generate_proof(tree_index).unwrap()
-            });
-        });
+    let k = u8::BITS - max_thread_count.leading_zeros() - 1; // floor(log_2(max_thread_count))
+    let lower_bound = 1u8 << k; // 2^k
+    let upper_bound = if k == 7 { u8::MAX } else { 1u8 << k + 1 }; // 2^(k+1)
+    let step = 1usize << k - 2; // 2^(k-2)
+
+    for i in (lower_bound..upper_bound).step_by(step) {
+        thread_counts.push(i);
     }
 
-    group.finish();
-}
+    thread_counts.push(max_thread_count);
 
-fn verify_proof(c: &mut Criterion) {
-    let mut group = c.benchmark_group("verify");
-    group.sample_size(10);
+    let mut ndm_smt = Option::<NdmSmt>::None;
+
+    for h in TREE_HEIGHTS.into_iter() {
+        for t in thread_counts.iter() {
+            if *t > MaxThreadCount::default().get_value() {
+                continue;
```

Comment:
When will this ever be reached? `thread_counts` always has a max value of `MaxThreadCount::default().get_value()`. Just remove the check since its superfluous

Reply:
It was more to pre-empt a possible error in the tuple construction, but I'll remove it